lib/Driver/ToolChains/Flang.cpp

```diff
@@ -600,27 +618,42 @@ void FlangFrontend::ConstructJob(Compilation &C, const JobAction &JA,
   // Add system include arguments.
   getToolChain().AddFlangSystemIncludeArgs(Args, UpperCmdArgs);
 
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("unix");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__unix");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__unix__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("linux");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__NO_MATH_INLINES");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LP64__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LONG_MAX__=9223372036854775807L");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__SIZE_TYPE__=unsigned long int");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__PTRDIFF_TYPE__=long int");
+  // Use clang's predefined macros
+  DiagnosticsEngine DE(new DiagnosticIDs(), new DiagnosticOptions, new IgnoringDiagConsumer());
```

Comment:
The first two parameters for the DiagnosticsEngine are of type `llvm::IntrusiveRefCntPtr`. The destructor for this class calls Release() which seems right [see here](https://llvm.org/doxygen/IntrusiveRefCntPtr_8h_source.html#l00156). However the third parameter is of type `DiagnosticConsumer *` and the `DiagnosticConsumer` class uses a default destructor.
https://github.com/flang-compiler/flang-driver/blob/9579fb8086134d4b9963a572b1b65df823169028/lib/Basic/Diagnostic.cpp#L433

This seems like a memory leak to me.

Reply:
Also, do you need to pass in the third argument? seems like it has a default value set to nullptr which should suffice. No?

Reply:
Rest of the changes look okay to me, can you rebase to have only one commit in your PR ?

Reply:
Constructor is,
```
  explicit DiagnosticsEngine(IntrusiveRefCntPtr<DiagnosticIDs> Diags,
                             DiagnosticOptions *DiagOpts,
                             DiagnosticConsumer *client = nullptr,
                             bool ShouldOwnClient = true);
```
and therefore the third argument `client` is managed by `DiagnosticsEnginer` and therefore there is no leak. The `nullptr` value may not suffice as there are methods where `client` is checked to be not ` nullptr` and if it is a `nullptr` an error occurs.

Reply:
Ok, that makes sense.